File: app/src/main/java/com/jimzrt/umsmounter/listadapters/ImageListAdapter.java
package com.jimzrt.umsmounter.listadapters;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.util.SortedList;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.ToggleButton;

import com.jimzrt.umsmounter.R;
import com.jimzrt.umsmounter.model.ImageItem;
import com.jimzrt.umsmounter.utils.Helper;

import java.util.List;

public class ImageListAdapter extends RecyclerView.Adapter<ImageListAdapter.ViewHolder> implements View.OnClickListener {

    private final Context mContext;
    private final OnImageListListener mCallback;
    private final SortedList<ImageItem> dataSet;
    private RecyclerView mRecyclerView;
    private int lastSelected = -1;

    public ImageListAdapter(List<ImageItem> data, Context context, OnImageListListener listener) {
        this.dataSet = new SortedList<>(ImageItem.class, new SortedList.Callback<ImageItem>() {
            @Override
            public int compare(ImageItem o1, ImageItem o2) {
                return o1.compareTo(o2);
            }

            @Override
            public void onChanged(int position, int count) {
                notifyItemRangeChanged(position, count);

            }

            @Override
            public boolean areContentsTheSame(ImageItem oldItem, ImageItem newItem) {
                return oldItem.equals(newItem);
            }

            @Override
            public boolean areItemsTheSame(ImageItem item1, ImageItem item2) {
                return item1.equals(item2);
            }

            @Override
            public void onInserted(int position, int count) {
                notifyItemRangeInserted(position, count);

            }

            @Override
            public void onRemoved(int position, int count) {
                notifyItemRangeRemoved(position, count);

            }

            @Override
            public void onMoved(int fromPosition, int toPosition) {
                notifyItemMoved(fromPosition, toPosition);

            }
        });

        if (data != null) {
            dataSet.beginBatchedUpdates();
            for (int i = 0; i < data.size(); i++) {
                dataSet.add(data.get(i));
            }
            dataSet.endBatchedUpdates();
        }


        this.mContext = context;
        this.mCallback = listener;

    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);

        mRecyclerView = recyclerView;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        // create a new view
        ConstraintLayout rootView = (ConstraintLayout) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.row_item, parent, false);

        ViewHolder holder = new ViewHolder(rootView, this);

        holder.imageView = rootView.findViewById(R.id.item_info);
        holder.name = rootView.findViewById(R.id.name);
        holder.size = rootView.findViewById(R.id.type);
        holder.mountButton = rootView.findViewById(R.id.mountButton);
        holder.mountButton.setOnClickListener(v -> mCallback.onMountImageButtonClicked());
        holder.deleteButton = rootView.findViewById(R.id.deleteButton);
        holder.deleteButton.setOnClickListener(v -> mCallback.onDeleteImageButtonClicked());
        holder.progressBar = rootView.findViewById(R.id.progressBar);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position, @NonNull List<Object> payloads) {
        if (payloads.isEmpty()) {
            // Perform a full update

            onBindViewHolder(holder, holder.getAdapterPosition());
        } else {
            // Perform a partial update
            for (Object payload : payloads) {
                View convertView = holder.rootView;

                if (payload.equals("aufklappen")) {

                    // boolean visible = ((LinearLayoutManager) mRecyclerView.getLayoutManager()).isViewPartiallyVisible(convertView, false, true);


                    // int colorFrom = Color.argb(0, 250, 250, 250);
                    // int colorTo = Color.argb(255, 220, 220, 220);
                    ValueAnimator colorAnimation = ValueAnimator.ofInt(0, 255).setDuration(1000);

                    colorAnimation.addUpdateListener(animator -> convertView.setBackgroundColor(Color.argb((Integer) colorAnimation.getAnimatedValue(), 220, 220, 220)));
                    colorAnimation.start();

                    // convertView.animate().rotationX(20f).setInterpolator(new AccelerateDecelerateInterpolator()).setDuration(500);

                    ValueAnimator anim0 = ValueAnimator.ofFloat(0, 20);
                    anim0.addUpdateListener(valueAnimator -> {
                        float val = (Float) valueAnimator.getAnimatedValue();
                        convertView.setRotationX(val);
                    });
                    anim0.setDuration(500);
                    anim0.start();

                    ValueAnimator anim = ValueAnimator.ofInt(Helper.px2dp(70, mContext), Helper.px2dp(70 + 30, mContext));
                    anim.addUpdateListener(valueAnimator -> {
                        int val = (Integer) valueAnimator.getAnimatedValue();
                        ViewGroup.LayoutParams layoutParams = convertView.getLayoutParams();
                        layoutParams.height = val;
                        convertView.setLayoutParams(layoutParams);
                    });
                    anim.setDuration(300);
                    anim.start();


                    convertView.postDelayed(() -> {
                        if (dataSet.get(holder.getAdapterPosition()).isDownloading()) {
                            holder.mountButton.setVisibility(View.INVISIBLE);
                        } else {
                            holder.mountButton.setVisibility(View.VISIBLE);

                        }
                        holder.deleteButton.setVisibility(View.VISIBLE);

                        mRecyclerView.smoothScrollToPosition(holder.getAdapterPosition());

                    }, 200);



                    // holder.deleteButton.setVisibility(View.VISIBLE);

                } else if (payload.equals("zuklappen")) {

                    ValueAnimator anim0 = ValueAnimator.ofFloat(convertView.getRotationX(), 0);
                    anim0.addUpdateListener(valueAnimator -> {
                        float val = (Float) valueAnimator.getAnimatedValue();
                        convertView.setRotationX(val);
                    });
                    anim0.setDuration(500);
                    anim0.start();

                    //  convertView.animate().rotationX(0).setInterpolator(new AccelerateDecelerateInterpolator()).setDuration(500).withEndAction(() -> {

                    //    });;

                    ValueAnimator colorAnimation = ValueAnimator.ofInt(convertView.getBackground().getAlpha(), 0).setDuration(1000);

                    colorAnimation.addUpdateListener(animator -> convertView.setBackgroundColor(Color.argb((Integer) colorAnimation.getAnimatedValue(), 220, 220, 220)));
                    colorAnimation.start();


                    ValueAnimator anim = ValueAnimator.ofInt(convertView.getLayoutParams().height, Helper.px2dp(70, mContext));
                    anim.addUpdateListener(valueAnimator -> {
                        int val = (Integer) valueAnimator.getAnimatedValue();
                        ViewGroup.LayoutParams layoutParams = convertView.getLayoutParams();
                        layoutParams.height = val;
                        convertView.setLayoutParams(layoutParams);
                    });

                    anim.setDuration(300);
                    anim.start();

                    convertView.postDelayed(() -> {
                        holder.mountButton.setVisibility(View.INVISIBLE);
                        holder.deleteButton.setVisibility(View.INVISIBLE);
                    }, 200);


                } else if (payload.equals("mount")) {

                    holder.imageView.setColorFilter(mContext.getResources().getColor(R.color.green));
                    holder.mountButton.setChecked(true);


                } else if (payload.equals("unmount")) {

                    holder.imageView.setColorFilter(null);
                    holder.mountButton.setChecked(false);
                } else if (payload.equals("download")) {
                    ImageItem imageItem = dataSet.get(holder.getAdapterPosition());
                    if (imageItem.isDownloading()) {
                        holder.progressBar.setVisibility(View.VISIBLE);
                        holder.progressBar.setProgress(imageItem.getProgress());
                        holder.size.setText(imageItem.getSize());
                    } else {
                        holder.progressBar.setVisibility(View.INVISIBLE);
                        holder.size.setText(imageItem.getSize());
                    }

                }
            }
        }
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        ImageItem imageItem = dataSet.get(position);


        holder.name.setText(imageItem.getName());
        holder.size.setText(imageItem.getSize());

        View convertView = holder.rootView;

        ViewGroup.LayoutParams layoutParams = convertView.getLayoutParams();
        if (imageItem.getSelected()) {


            convertView.setBackgroundColor(Color.argb(255, 220, 220, 220));
            convertView.setRotationX(20f);
            layoutParams.height = Helper.px2dp(70 + 30, mContext);

            if (!imageItem.isDownloading()) {
                holder.mountButton.setVisibility(View.INVISIBLE);

            } else {
                holder.mountButton.setVisibility(View.VISIBLE);

            }
            holder.deleteButton.setVisibility(View.VISIBLE);
        } else {
            convertView.setBackgroundColor(Color.argb(0, 220, 220, 220));
            convertView.setRotationX(0);
            layoutParams.height = Helper.px2dp(70, mContext);
            holder.mountButton.setVisibility(View.INVISIBLE);
            holder.deleteButton.setVisibility(View.INVISIBLE);
        }


        convertView.setLayoutParams(layoutParams);
        //holder.mountButton.setAlpha(0);
        //holder.deleteButton.setAlpha(0);


        //  }
        //  } else {
        //      bgColor = Color.WHITE;
        //  }

        //   if(imageItem.getMounted()){
        //       holder.imageView.setColorFilter(mContext.getResources().getColor(R.color.green));
        //       holder.mountButton.setChecked(true);

        //   } else {
        //  holder.imageView.setColorFilter(null);
        //    holder.mountButton.setChecked(false);

        //    }

        if (imageItem.isDownloading()) {
            holder.progressBar.setVisibility(View.VISIBLE);
            holder.progressBar.setProgress(imageItem.getProgress());
        } else {
            holder.progressBar.setVisibility(View.INVISIBLE);
        }


    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }


    public void remove(ImageItem checkedItem) {
        //animationMap.remove(dataSet.indexOf(checkedItem));
        //   setSelectedItemPosition(-1);
        dataSet.remove(checkedItem);
    }

    @Override
    public void onClick(View v) {
        int itemPosition = mRecyclerView.getChildLayoutPosition(v);

        //setSelectedItemPosition(itemPosition);
        ImageItem item = dataSet.get(itemPosition);
        mCallback.onImageSelected(item);

    }

    public void addItems(List<ImageItem> list) {

        // dataSet.beginBatchedUpdates();
        //   dataSet.clear();

        //    dataSet.beginBatchedUpdates();
        for (int j = 0; j < dataSet.size(); j++) {
            if (!list.contains(dataSet.get(j))) {
                dataSet.remove(dataSet.get(j));
            }
        }
        //     if(dataSet.size() == 0){

        for (int i = 0; i < list.size(); i++) {
            //  addItem(list.get(i), false);
            if (!contains(list.get(i))) {
                dataSet.add(list.get(i));
            }
        }
//        }


        // dataSet.endBatchedUpdates();

    }

    public int addItem(ImageItem imageItem) {

        return dataSet.add(imageItem);

    }

    private boolean contains(ImageItem item) {
        return dataSet.indexOf(item) != -1;
    }

    public int getPositionOfItem(ImageItem imageItem) {
        return dataSet.indexOf(imageItem);
    }

    public void setSelectedItem(ImageItem selectedItem) {
        if (selectedItem == null) {
            setSelectedItemPosition(-1);
        } else {
            int position = dataSet.indexOf(selectedItem);
            setSelectedItemPosition(position);
        }

    }


    private void setSelectedItemPosition(int position) {
        //notifyDataSetChanged();
        if (position == -1) {
            if (lastSelected == -1) {
                return;
            }
            dataSet.get(lastSelected).setSelected(false);
            notifyItemChanged(lastSelected, "zuklappen");
            lastSelected = -1;
            return;
        }

        if (lastSelected == position) {
            dataSet.get(position).setSelected(false);
            notifyItemChanged(position, "zuklappen");
            lastSelected = -1;
            return;
        } else if (lastSelected != -1) {
            dataSet.get(lastSelected).setSelected(false);
            notifyItemChanged(lastSelected, "zuklappen");

        }
        dataSet.get(position).setSelected(true);
        notifyItemChanged(position, "aufklappen");
        lastSelected = position;
    }


    public interface OnImageListListener {

        void onMountImageButtonClicked();

        void onDeleteImageButtonClicked();

        void onImageSelected(ImageItem item);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final ConstraintLayout rootView;
        private ImageView imageView;
        private TextView name;
        private TextView size;
        private ToggleButton mountButton;
        private Button deleteButton;
        private ProgressBar progressBar;

        ViewHolder(View rootView, View.OnClickListener listener) {
            super(rootView);
            this.rootView = (ConstraintLayout) rootView;
            this.rootView.setOnClickListener(listener);
        }
    }
}

File: app/src/main/java/com/jimzrt/umsmounter/listadapters/ImageListAdapter.kt
package com.jimzrt.umsmounter.listadapters

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SortedList
import com.jimzrt.umsmounter.R
import com.jimzrt.umsmounter.model.ImageItem
import com.jimzrt.umsmounter.utils.Helper.px2dp

class ImageListAdapter(data: List<ImageItem>?, context: Context, listener: OnImageListListener) : RecyclerView.Adapter<ImageListAdapter.ViewHolder>(), View.OnClickListener {
    private val mContext: Context
    private val mCallback: OnImageListListener
    private val dataSet: SortedList<ImageItem>
    private var mRecyclerView: RecyclerView? = null
    private var lastSelected = -1
    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        mRecyclerView = recyclerView
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        // create a new view
        val rootView = LayoutInflater.from(parent.context)
                .inflate(R.layout.row_item, parent, false) as ConstraintLayout
        val holder = ViewHolder(rootView, this)
        holder.imageView = rootView.findViewById(R.id.item_info)
        holder.name = rootView.findViewById(R.id.name)
        holder.size = rootView.findViewById(R.id.type)
        holder.mountButton = rootView.findViewById(R.id.mountButton)
        holder.mountButton?.setOnClickListener { mCallback.onMountImageButtonClicked() }
        holder.deleteButton = rootView.findViewById(R.id.deleteButton)
        holder.deleteButton?.setOnClickListener { mCallback.onDeleteImageButtonClicked() }
        holder.progressBar = rootView.findViewById(R.id.progressBar)
        return holder
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            // Perform a full update
            onBindViewHolder(holder, holder.adapterPosition)
        } else {
            // Perform a partial update
            for (payload in payloads) {
                val convertView: View = holder.rootView
                when (payload) {
                    "aufklappen" -> {

                        // boolean visible = ((LinearLayoutManager) mRecyclerView.getLayoutManager()).isViewPartiallyVisible(convertView, false, true);


                        // int colorFrom = Color.argb(0, 250, 250, 250);
                        // int colorTo = Color.argb(255, 220, 220, 220);
                        val colorAnimation = ValueAnimator.ofInt(0, 255).setDuration(1000)
                        colorAnimation.addUpdateListener { convertView.setBackgroundColor(Color.argb((colorAnimation.animatedValue as Int), 235, 235, 235)) }
                        colorAnimation.start()

                        // convertView.animate().rotationX(20f).setInterpolator(new AccelerateDecelerateInterpolator()).setDuration(500);
                        val anim0 = ValueAnimator.ofFloat(0f, 15f)
                        anim0.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Float
                            convertView.rotationX = `val`
                        }
                        anim0.duration = 500
                        anim0.start()
                        val anim = ValueAnimator.ofInt(px2dp(70, mContext), px2dp(70 + 70, mContext))
                        anim.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Int
                            val layoutParams = convertView.layoutParams
                            layoutParams.height = `val`
                            convertView.layoutParams = layoutParams
                        }
                        anim.duration = 300
                        anim.start()
                        convertView.postDelayed({
                            if (dataSet[holder.adapterPosition].isDownloading) {
                                holder.mountButton!!.visibility = View.INVISIBLE
                            } else {
                                holder.mountButton!!.visibility = View.VISIBLE
                            }
                            holder.deleteButton!!.visibility = View.VISIBLE
                            mRecyclerView!!.smoothScrollToPosition(holder.adapterPosition)
                        }, 230)


                        // holder.deleteButton.setVisibility(View.VISIBLE);
                    }
                    "zuklappen" -> {
                        val anim0 = ValueAnimator.ofFloat(convertView.rotationX, 0f)
                        anim0.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Float
                            convertView.rotationX = `val`
                        }
                        anim0.duration = 500
                        anim0.start()

                        //  convertView.animate().rotationX(0).setInterpolator(new AccelerateDecelerateInterpolator()).setDuration(500).withEndAction(() -> {

                        //    });;
                        val colorAnimation = ValueAnimator.ofInt(convertView.background.alpha, 0).setDuration(1000)
                        colorAnimation.addUpdateListener { convertView.setBackgroundColor(Color.argb((colorAnimation.animatedValue as Int), 235, 235, 235)) }
                        colorAnimation.start()
                        val anim = ValueAnimator.ofInt(convertView.layoutParams.height, px2dp(70, mContext))
                        anim.addUpdateListener { valueAnimator: ValueAnimator ->
                            val `val` = valueAnimator.animatedValue as Int
                            val layoutParams = convertView.layoutParams
                            layoutParams.height = `val`
                            convertView.layoutParams = layoutParams
                        }
                        anim.duration = 300
                        anim.start()
                        convertView.postDelayed({
                            holder.mountButton!!.visibility = View.INVISIBLE
                            holder.deleteButton!!.visibility = View.INVISIBLE
                        }, 130)
                    }
                    "mount" -> {
                        holder.imageView!!.setColorFilter(mContext.resources.getColor(R.color.green))
                        holder.mountButton!!.isChecked = true
                    }
                    "unmount" -> {
                        holder.imageView?.colorFilter = null
                        holder.mountButton!!.isChecked = false
                    }
                    "download" -> {
                        val imageItem = dataSet[holder.adapterPosition]
                        if (imageItem.isDownloading) {
                            holder.progressBar!!.visibility = View.VISIBLE
                            holder.progressBar?.progress = imageItem.progress
                        } else {
                            holder.progressBar!!.visibility = View.INVISIBLE
                        }
                        holder.size!!.text = imageItem.size
                    }
                }
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val imageItem = dataSet[position]
        holder.name!!.text = imageItem.name
        holder.size!!.text = imageItem.size
        val convertView: View = holder.rootView
        val layoutParams = convertView.layoutParams
        if (imageItem.selected) {
            convertView.setBackgroundColor(Color.argb(255, 220, 220, 220))
            convertView.rotationX = 20f
            layoutParams.height = px2dp(70 + 30, mContext)
            if (!imageItem.isDownloading) {
                holder.mountButton!!.visibility = View.INVISIBLE
            } else {
                holder.mountButton!!.visibility = View.VISIBLE
            }
            holder.deleteButton!!.visibility = View.VISIBLE
        } else {
            convertView.setBackgroundColor(Color.argb(0, 220, 220, 220))
            convertView.rotationX = 0f
            layoutParams.height = px2dp(70, mContext)
            holder.mountButton!!.visibility = View.INVISIBLE
            holder.deleteButton!!.visibility = View.INVISIBLE
        }
        convertView.layoutParams = layoutParams
        //holder.mountButton.setAlpha(0);
        //holder.deleteButton.setAlpha(0);


        //  }
        //  } else {
        //      bgColor = Color.WHITE;
        //  }

        //   if(imageItem.getMounted()){
        //       holder.imageView.setColorFilter(mContext.getResources().getColor(R.color.green));
        //       holder.mountButton.setChecked(true);

        //   } else {
        //  holder.imageView.setColorFilter(null);
        //    holder.mountButton.setChecked(false);

        //    }
        if (imageItem.isDownloading) {
            holder.progressBar!!.visibility = View.VISIBLE
            holder.progressBar?.progress = imageItem.progress
        } else {
            holder.progressBar!!.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        return dataSet.size()
    }

    fun remove(checkedItem: ImageItem) {
        //animationMap.remove(dataSet.indexOf(checkedItem));
        //   setSelectedItemPosition(-1);
        dataSet.remove(checkedItem)
    }

    override fun onClick(v: View) {
        val itemPosition = mRecyclerView!!.getChildLayoutPosition(v)

        //setSelectedItemPosition(itemPosition);
        val item = dataSet[itemPosition]
        mCallback.onImageSelected(item)
    }

    fun addItems(list: List<ImageItem>) {

        // dataSet.beginBatchedUpdates();
        //   dataSet.clear();

        //    dataSet.beginBatchedUpdates();
        for (j in 0 until dataSet.size()) {
            if (!list.contains(dataSet[j])) {
                dataSet.remove(dataSet[j])
            }
        }
        //     if(dataSet.size() == 0){
        for (i in list.indices) {
            //  addItem(list.get(i), false);
            if (!contains(list[i])) {
                dataSet.add(list[i])
            }
        }
        //        }


        // dataSet.endBatchedUpdates();
    }

    fun addItem(imageItem: ImageItem): Int {
        return dataSet.add(imageItem)
    }

    private operator fun contains(item: ImageItem): Boolean {
        return dataSet.indexOf(item) != -1
    }

    fun getPositionOfItem(imageItem: ImageItem): Int {
        return dataSet.indexOf(imageItem)
    }

    fun setSelectedItem(selectedItem: ImageItem?) {
        if (selectedItem == null) {
            setSelectedItemPosition(-1)
        } else {
            val position = dataSet.indexOf(selectedItem)
            setSelectedItemPosition(position)
        }
    }

    private fun setSelectedItemPosition(position: Int) {
        //notifyDataSetChanged();
        if (position == -1) {
            if (lastSelected == -1) {
                return
            }
            dataSet[lastSelected].selected = false
            notifyItemChanged(lastSelected, "zuklappen")
            lastSelected = -1
            return
        }
        if (lastSelected == position) {
            dataSet[position].selected = false
            notifyItemChanged(position, "zuklappen")
            lastSelected = -1
            return
        } else if (lastSelected != -1) {
            dataSet[lastSelected].selected = false
            notifyItemChanged(lastSelected, "zuklappen")
        }
        dataSet[position].selected = true
        notifyItemChanged(position, "aufklappen")
        lastSelected = position
    }

    interface OnImageListListener {
        fun onMountImageButtonClicked()
        fun onDeleteImageButtonClicked()
        fun onImageSelected(item: ImageItem?)
    }

    class ViewHolder(rootView: View, listener: View.OnClickListener?) : RecyclerView.ViewHolder(rootView) {
        val rootView: ConstraintLayout = rootView as ConstraintLayout
        var imageView: ImageView? = null
        var name: TextView? = null
        var size: TextView? = null
        var mountButton: ToggleButton? = null
        var deleteButton: Button? = null
        var progressBar: ProgressBar? = null

        init {
            this.rootView.setOnClickListener(listener)
        }
    }

    init {
        dataSet = SortedList(ImageItem::class.java, object : SortedList.Callback<ImageItem>() {
            override fun compare(o1: ImageItem, o2: ImageItem): Int {
                return o1.compareTo(o2)
            }

            override fun onChanged(position: Int, count: Int) {
                notifyItemRangeChanged(position, count)
            }

            override fun areContentsTheSame(oldItem: ImageItem, newItem: ImageItem): Boolean {
                return oldItem == newItem
            }

            override fun areItemsTheSame(item1: ImageItem, item2: ImageItem): Boolean {
                return item1 == item2
            }

            override fun onInserted(position: Int, count: Int) {
                notifyItemRangeInserted(position, count)
            }

            override fun onRemoved(position: Int, count: Int) {
                notifyItemRangeRemoved(position, count)
            }

            override fun onMoved(fromPosition: Int, toPosition: Int) {
                notifyItemMoved(fromPosition, toPosition)
            }
        })
        if (data != null) {
            dataSet.beginBatchedUpdates()
            for (i in data.indices) {
                dataSet.add(data[i])
            }
            dataSet.endBatchedUpdates()
        }
        mContext = context
        mCallback = listener
    }
}